test/core/iomgr/pollset_windows_starvation_test.cc

```diff
@@ -0,0 +1,119 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#if defined(GRPC_WINSOCK_SOCKET)
+
+#include <vector>
+
+#include <grpc/grpc.h>
+#include <grpc/support/time.h>
+
+#include "src/core/lib/gprpp/thd.h"
+#include "src/core/lib/iomgr/exec_ctx.h"
+#include "src/core/lib/iomgr/iocp_windows.h"
+#include "src/core/lib/iomgr/iomgr_internal.h"
+#include "src/core/lib/iomgr/pollset.h"
+#include "src/core/lib/iomgr/pollset_windows.h"
+#include "src/core/lib/surface/init.h"
+#include "test/core/util/test_config.h"
+
+struct ThreadParams {
+  gpr_cv cv;
+  gpr_mu mu;
+  int complete;
+};
+
+int main(int argc, char** argv) {
+  grpc_init();
+
+  // Create three threads that all start queueing for work.
+  //
+  // The first one becomes the active poller for work and the two other
+  // threads go into the poller queue.
+  //
+  // When work arrives, the first one notifies the next active poller,
+  // this wakes the second thread - however all this does is return from
```

Comment:
I do not think the description here is accurate. All these three threads are using different pollsets, so both the second and third threads would get stuck without your fix. Basically after patching pollset_kick, the second thread is waken up by the first thread through g_global_root_worker and the third thread is waken up by the second thread.

Reply:
This is the active poller wait segment:

https://github.com/grpc/grpc/blob/a8ab03d7587bc86361b61ba1d91ba619af285d50/src/core/lib/iomgr/pollset_windows.cc#L124-L154

> The first one becomes the active poller for work

Line 130 is where the active poller waits for more work. The if-statement is guarded with `if (g_active_poller == NULL)` and few lines later sets the `g_active_poller = &worker` to ensure there's only one active poller running code inside that if-statement.

> When work arrives, the first one notifies the next ~~active~~ queued poller, this wakes the second thread.

After the first thread receives work and exits the `grpc_iocp_work` call on L130, it will proceed to L136..146

https://github.com/grpc/grpc/blob/a8ab03d7587bc86361b61ba1d91ba619af285d50/src/core/lib/iomgr/pollset_windows.cc#L136-L146

This bit of code tries to acquire another queued poller (first through the current pollset and then through the global queue). If such poller is available, it gets kicked.

This ensures the _second_ thread will be waken up.

> It's up to that thread to figure out if it still wants to queue for more work or if it should kick
 other pollers.

The above code will only ever kick one additional poller. If the second thread queues for more work, it will then kick the third thread once it receives work. However if the second thread decides it's not needed and wants to terminate, it ends up notifying the pollsets through the `pollset_kick` function, which did not have the same "local or global" functionality as the `pollset_work`.

------------

Save for "next active poller" vs "next queued poller" issue, I don't think there's anything inaccurate in the comment.

Reply:
Ah, thanks for the detailed explanation! I had not realized this in the first place.